오픈마켓 II [STEP 1] 주디, 재재

ZZBAE opened the pull request:
@protocorn93 콘 안녕하세요!
오픈마켓II의 step 1-1, 1-2 (브랜치의 step3-1, 3-2) PR 보냅니다.
리뷰 잘 부탁드리겠습니다 🙇🏻‍♀️


# 🛍 **오픈마켓II "OPEN MARKET II"**
> **프로젝트 기간** : 2022.07.21 목  ~ 2022.07.29 금 </br>**팀원** : 주디, 재재</br>**리뷰어** : 콘

## 💾 **개발환경 및 라이브러리**
[![swift](https://img.shields.io/badge/swift-5.6-orange)]()
[![xcode](https://img.shields.io/badge/Xcode-13.4.1-blue)]()

---

## 💡 **Step 1-1, 1-2 수행 중 핵심경험**
- [x] multipart/form-data의 구조 파악
- [x] URLSession을 활용한 multipart/form-data 요청 전송
- [x] 사용자 친화적인 UI/UX 구현 (적절한 입력 컴포넌트 사용, 알맞은 키보드 타입 지정)
- [x] UIAlertController 액션의 completion handler 활용
- [x] UIAlertController의 textFields 활용
- [x] UICollectionView 를 통한 좌우 스크롤 기능 구현
<br>

## 📱 **구현한 화면**

| 상품의 ListView | 상품의 GridView | 상품을 등록하는 화면 |
|:---:|:---:|:---:|
| ![](https://i.imgur.com/qODkdGJ.gif) | ![](https://i.imgur.com/LXZsYu6.gif) | ![](https://i.imgur.com/djsjE9H.gif) |

| 상품 기본 정보 입력 | 상품 상세 설명 입력 | 상품의 이미지 최대 5장까지 추가 |
|:---:|:---:|:---:|
|![](https://i.imgur.com/YjFMx7A.gif)|![](https://i.imgur.com/AOUbqYN.gif)|![Simulator Screen Recording - iPhone 13 Pro - 2022-07-29 at 16 43 30](https://user-images.githubusercontent.com/98302604/181710105-bdfd70f7-32db-4aa2-82a8-bfdc6974efc4.gif)|
|상품명, 상품가격, 할인가격, 재고수량 정보 입력할 때 해당 키보드가 나오게끔 구현|상품의 상세 설명이 길어져도 밑에서 올라오는 키보드가 설명칸을 가리지 않게끔 구현|상품 이미지를 정사각형으로 넣을 수 있고, 1장 ~ 5장을 추가할 수 있도록 구현|


<br>

---

## 🧑‍💻 **전반적인 코드 설명**

### **Model** 
#### 1) **Param**
> 상품 등록 화면에 표시되는 해당 상품 정보들의 타입
- `vendorInfo`: 판매자의 `secret`과 `identifier`를 담은 열거형
- `Param`: productName(상품명), price(상품가격), discountedPrice(할인가격), currency(통화단위), stock(재고수량), description(상품상세설명), secret(VendorInfo의 사용자 비밀번호)
- `ImageParam`: imageName(이미지 파일 이름), imageType(이미지 파일 타입), imageData(이미지 데이터 타입)
<br>
### **View**

#### 화면 구성 정리본

![](https://i.imgur.com/uo6MJEY.png)

#### 1) AddProductView
> 새로운 상품을 등록(또는 수정)하는 화면
- `arrangeSubView`: 요소들을 `SubView`에 넣고 제약을 설정하는 메서드 
- `createParam`: 입력된 정보들을 `Param`으로 반환해주는 메서드
- `configure`: `Param`의 각 요소들을 각각의 칸`(TextField)`에 넣어주는 메서드


#### 2) AddProductCollectiongViewCell
> 상품 등록 화면에서 이미지의 설정을 해주는 셀
- `arrangeSubView`: 상품 이미지의 constraint를 설정해주는 메서드

<br>

### **ViewController**
#### 1) **AddProductViewController**
> 
- `changeToEditMode`: 상품등록화면을 상품수정화면에서 사용할 수 있게끔 해주는 메서드
- `configureUI`: 상단의 네비게이션 아이템 넣어준 부분의 layout 잡아주는 메서드
- `goBack`: 상품리스트 화면으로 돌아가는 objc 메서드
- `goBackWithUpdate`: 상품의 정보들을 입력해준 후 post 해주고, 성공과 실패의 얼럿을 띄워주는 메서드
- `CollectionView's DataSource & Delegate`: `UICollectionViewDataSource`, `UICollectionViewDelegate`를 채택하여 뷰를 재사용할 수 있고, 이미지의 indexPath를 계산하여 얼럿을 띄워주고 제약을 걸어주는 메서드
- `ImagePickerController`: `UIImagePickerControllerDelegate`와 `UINavigationControllerDelegate`를 채택하여, post할 이미지의 설정과 제약을 걸어주는 메서드
- `UITextView`: `textView`에 작성을 시작하고 끝낼때의 `viewConstraint`를 조절해주는 메서드
<br>

---

## ⁉️ **궁금한 점**
### 1. 텍스트뷰 자동 스크롤 - 키보드 사이즈
**TextView**에 글자 입력 시 글자가 키보드에 가려지지 않게 즉, 키보드 위로 자동 스크롤이 되도록 하는 요구사항이 있습니다. 키보드 위로 글자가 보이게 하는 방법으로 **TextView**의 크기를 줄였다가 다시 늘리는 방법으로 구현해봤습니다.

#### StackView의 Constraint 변경
```swift
extension ProductViewController: UITextViewDelegate {
    func textViewDidBeginEditing(_ textView: UITextView) {
        productView.entireStackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -300).isActive = true
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
        productView.entireStackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -8).isActive = true
    }
}
```
**TextView**가 속해있는 **StackView**의 **Constraint**를 변경해보려고 했습니다. 키보드 크기만큼 **bottom**을 올렸다가 입력이 끝나면 내리도록 했는데, 올리는 것은 성공했으나 다시 내려오지 않았습니다. 원인은 계속 새로운 **Constraint**를 추가해서 제약이 상충됐기 때문이었습니다. 

```swift
private lazy var viewConstraint = productView.entireStackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -260)

extension AddProductViewController: UITextViewDelegate {
    func textViewDidBeginEditing(_ textView: UITextView) {
        viewConstraint.isActive = true
    }
    
    func textViewDidEndEditing(_ textView: UITextView) {
        viewConstraint.isActive = false
    }
}
```
하나의 **Constraint**를 선언해 `isActive`를 조절하는 방식으로 변경하여 해결했습니다. 또한 기본적으로 뷰를 구성할 때 지정한 **bottom** 제약의 **Priority**를 낮춰서 상충없이 동작하도록 구현했습니다.

하지만 현재 `260`과 같이 키보드의 크기를 숫자로 맞춰서 조절하는 상태인데 혹시 키보드의 사이즈를 구해 정확하게 조절이 필요할까요? 그렇다면 어떻게 구할 수 있는지, 혹은 이렇게만 해도 요구사항을 충족하니 괜찮은 방법인지 궁금합니다 🤔
<br>

### 2. 사진을 5개 넣으면 POST 되지 않는 문제
사진은 최소 1장에서 최대 5장까지 가능하도록 요구사항이 있습니다. 하지만 4장까지는 문제없이 **post**가 되는데 5장 올리려고 하면 실패하고 400번대 응답이 돌아옵니다.. 

<img src="https://i.imgur.com/lcyZR5A.png" width="500" height="600" />
<br><br>

**post**하려는 사진 배열을 출력했을 때 데이터는 잘 들어있는 것을 위 사진처럼 확인했는데 어떤 점이 원인인지 모르겠습니다 😢


### 3. 사진 타입 구하기
상품 정보를 **post**할 때 이미지 정보를 아래와 같이 넣어주고 있습니다.
```swift
"key": "images",
"src": "사진이름.jpeg",
"type": "jpeg"
```
사용자가 앨범에서 선택한 사진을 `UIImage` 타입으로 받아 **post**하는데 `UIImage`로 받은 파일의 타입을 알 수 있는 방법이 있을까요? 🤔

현재는 타입을 `"jpeg"`를 기본값으로 지정해 모든 사진이 `jpeg`로 등록되는데 문제없이 **post**는 되고 있습니다!
<br>

### 4. 사진 추가 셀의 이미지 (+)

| 현재 이미지 | 요구사항 이미지 | 
| :--------: | :--------: | 
| <img src="https://i.imgur.com/EkrQaIC.png)" width="100" height="100" />| <img src="https://i.imgur.com/562Ilvx.png" width="100" height="100" />|


사진을 추가하는 버튼 역할을 하는 **Cell**의 이미지는 `UIImage(systemName: "plus")`를 사용하고 있습니다. 요구사항처럼 작게 표시하려면 `imageView.contentMode = .center` 처럼 `contentMode`를 변경하면 가능은 하지만 추가한 사진들의 크기가 조절되지 않아서 일단 그대로 두었습니다. 동작하는 데는 상관은 없지만 혹시 요구사항처럼 표현할 수 있는 다른 방법이 있을지 궁금합니다.
<br>

## 고민한 점 
### 1. 스크롤 시 상품에 맞지 않는 이미지가 나오는 문제 해결
빠르게 스크롤 하다보면 엉뚱한 사진이 띄워져 있는 문제가 있었습니다. 서버를 통해 사진을 받아오는 요청은 비동기로 진행되고, **Cell**이 재사용되기 때문에 발생하는 문제였습니다. 사진의 용량이 크거나 느린 네트워크일 경우 이전에 요청한 사진이 뒤늦게 돌아와서 잘못된 이미지가 떴습니다.

[**예시 - 제다이가 된 주디와 웡빙 (원래는 연예인 사진)**]
<img src="https://i.imgur.com/TnUm6tg.jpg" width="250" height="500" />

<br>이미지를 요청한 **Cell**과 현재 **Cell**이 일치하지 않았을 때의 문제이므로 **Cell**에 이미지를 할당하기 전에 비교하여 **Cell**이 일치한 경우에만 이미지를 넣도록 하여 해결했습니다.

```swift
 guard indexPath == self.collectionView.indexPath(for: cell) else { return }
```
<br>

### 2. 상위 셀 생성해 상속받기
이전 스텝에서 사용한 **LIST**와 **Grid Cell**이 중복된 코드가 많아 해결하고 싶었습니다.
`UICollectionViewCell`을 상속받는 `MarketCollectionViewCell`을 생성해 중복되는 UI와 메서드를 갖도록 했습니다. 이후 **LIST**와 **Grid Cell**이 `MarketCollectionViewCell`를 상속받아 사용할 수 있도록 하여 코드의 중복을 줄이고 추상화해 줄 수 있었습니다.
<br>

### 3. Alert은 Main Thread에서 실행하기
**get**이나 **post**와 같은 서버 요청이 실패한 경우 얼럿을 띄워 사용자에게 알려주도록 했는데 통신이 실패했을 때 얼럿이 뜨지 않고 아래와 같은 에러가 발생했습니다.

**Alert을 메인 스레드에서 하지 않았을 때 나타난 에러**
![](https://i.imgur.com/IkZy5AD.png)

얼럿을 띄우는 것 역시 UI 동작이기 때문에 `Main Thread`에서 실행되도록 변경해 해결했습니다!

ZZBAE committed:
refactor: 필요없는 부분 주석 제거 및 함수 정리

ZZBAE committed:
feat: 데이터와 이미지 post 하는 연습용 코드 생성

Judy-999 committed:
refactor: extension을 통해 강제 언래핑 제거

- String을 extension해 convertData 연산 프로퍼티로 Data 타입으로 전환
- Key와 Value가 String인 Dictionary를 extension해 bringValue 함수로 Value를 가져옴
- 통신 실패 시 Alert을 띄우는 코드를 main 스레드에서 하도록 수정

Judy-999 committed:
feat: 상품 데이터를 수정하는 patchData 연습용 코드 생성

ZZBAE committed:
feat: 상품 데이터를 제거하는 deleteData 연습용 코드 생성

- delete를 하기 위한 상품 secretKey를 response하는 inquireSecretKey 메서드 구현

Judy-999 committed:
refactor: List와 Grid Cell의 상위 클래스인 MarketCollectionViewCell 생성

- 중복된 뷰와 메서드를 상위 클래스에서 구현한 후 상속을 통해 추상화

Judy-999 committed:
refactor: 셀의 이미지를 설정하는 configureImage 메서드를 생성

- 빠르게 스크롤 시 해당 셀의 상품과 맞지 않은 이미지가 들어가있는 현상 해결 및 수정

ZZBAE committed:
feat: 상품 등록 뷰에서 이미지 콜렉션 뷰와 뷰셀 생성

- AddProductCollectionViewCell: 상품의 이미지를 추가하는 수평 스크롤 뷰 생성 객체
- AddProductView: 상품을 등록하는 뷰

Judy-999 committed:
feat: 상품등록 뷰의 DataSource 구현

- 상품등록 뷰를 담당하는 AddProductViewController 생성

ZZBAE committed:
feat: AddProductView의 UI 추가 및 layout 설정

- AddProductView에서 상품등록화면을 구성하는 요소들의 layout 설정

Judy-999 committed:
feat: + 버튼에 상품등록 화면으로 이동하는 action 등록

- UIBarButtonItem 생성하는 방식 변경

ZZBAE committed:
feat: NavigationTitle과 BarButton 생성

- NavigationTitle을 "상품등록"으로 생성
- 기존 backButton은 hide 해주어 제거해줌
- leftBarButtonItem에 "Cancel" 버튼과 rightBarButtonItem에 "Done" 버튼 생성
- Cancel과 Done 클릭 시 MarketCollectionViewController로 이동 구현

Judy-999 committed:
feat: 사진을 가져오는 기능을 UIImagePickerController를 이용해 구현

- UIImagePickerController를 이용해 앨범에 있는 사진을 정사각형으로 편집해서 가져오도록 함
- AddProductCollectionViewCell의 productImageButton을 productImage로 변경
- 추가한 사진이 5개를 초과한 후 추가 셀을 누르면 추가 불가능을 알리는 얼럿을 띄우게 함

ZZBAE committed:
feat: 텍스트 입력 중 키보드가 콘텐츠를 가리지 않게 하는 제약 추가

- TextViewDelegate를 사용하여 키보드가 올라왔을 때 TextView를 가리지 않도록 구현

Judy-999 committed:
feat: Post할 데이터를 담는 Param, ImageParam 타입 생성

- AddProductView에 입력된 텍스트로 Param 타입을 생성하는 receiveParam 메서드 추가
- bargainPrice와 stock은 비어있으면 기본값 "0"으로 지정

Judy-999 committed:
feat: 새로운 상품을 등록하는 로직 구현

- 상품등록 화면에서 Done 버튼을 누르면 입력된 데이터를 Post 하는 코드 추가
- URLSessionManager의 임시 postData 메서드를 실제 통신 가능하도록 변경

ZZBAE committed:
feat: 상품등록의 부수적인 기능 구현 및 네이밍 변경

- bargainPrice를 discountedPrice로 네이밍 변경
- bargainPriceTextField를 discountedPriceTextField로 네이밍 변경
- 이미지가 2개 이상일 때 post 되지 않는 현상 해결
- 필수 항목을 입력하지 않았을 때 상품 등록이 불가한 얼럿 생성

Judy-999 committed:
feat: 이미지 용량을 줄이는 compressImage 메서드 구현

- ImageParam의 imageData를 UIImage에서 Data 타입으로 변경

ZZBAE committed:
feat: 상품수정모드가 가능하도록 changeToEditMode 메서드 구현

- 수정할 상품의 정보를 상품수정 뷰에 띄울 수 있게 해주는 configure 메서드 구현
- viewDidAppear를 이용하여 상품등록 및 수정 후 즉시 뷰에 나타날 수 있도록 수정
- 필요없는 주석 제거 및 //MARK: 로 메서드 구분

ZZBAE committed:
refactor: 타입 및 post 로직 수정

- price 타입을 Int 에서 Double로 수정
- post를 성공했을 때만 화면을 나갈 수 있도록 로직 변경

ZZBAE committed:
refactor: layout 오류 해결 및 text 설정 변경

- 각 textField에 맞는 키보드 타입 설정
- stackBottomConstraint의 priority 설정 (layout 오류 수정)
- price의 길이가 길어져도 보이게끔 전체 stackView의 alignment를 fill로 변경

Judy-999 committed:
refactor: 상품등록 성공 로직 변경

- 상품등록(post)이 성공하면 이전 화면으로 돌아가는 코드를 메인 스레드에서 실행

ZZBAE committed:
Update README.md

protocorn93 commented on OpenMarket/OpenMarket/Session/URLSessionManager.swift:
```diff
@@ -51,7 +51,114 @@ final class URLSessionManager {
+extension String {
+    fileprivate var convertData: Data {
+        guard let data = self.data(using: .utf8) else {
+            return Data()
```
변환에 실패했을 땐 `nil` 을 반환하는게 맥락에 맞지 않을까요??

ZZBAE replied:
옵셔널 값을 바인딩해주는 함수를 만들어주고 싶었기 때문에 만든 메서드인데, nil을 반환하면 바인딩이 되지 않아 임의로 Data()를 반환시켜주었습니다. 하지만 리뷰를 확인하고 저희가 처음 생각한 의도와 다른 것 같아 메서드를 전부 제거해주고 guard let 으로 수정해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/Session/URLSessionManager.swift:
```diff
@@ -51,7 +51,114 @@ final class URLSessionManager {
+        }
+        
+        return value as? UIImage ?? UIImage()
+    }
```
위와 마찬가지로 키에 해당하는 값이 없으면 `nil`을 반환해주는게 맥락에 맞지 않을까요?? `""`이나 `UIImage` 처럼 default 값을 반환해주신 이유가 궁금해요~

ZZBAE replied:
위의 코멘트에 달아놓았듯이 옵셔널 값을 바인딩해주는 함수를 만들어주고 싶었기 때문에 만든 메서드인데, nil을 반환하면 바인딩이 되지 않아 임의로 Data()를 반환시켜주었습니다. 하지만 리뷰를 확인하고 저희가 처음 생각한 의도와 다른 것 같아 메서드를 전부 제거해주고 guard let 으로 수정해주었습니다!

protocorn93 commented on OpenMarket/OpenMarket/View/AddProductCollectionViewCell.swift:
```diff
@@ -0,0 +1,43 @@
+final class AddProductCollectionViewCell: UICollectionViewCell {
+    static let id = "ProductCell"
+
+    let productImage: UIImageView = {
+        let imageView = UIImageView()
+        imageView.backgroundColor = .systemGray5
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
```
셀 안의 UI 요소들에도 접근제어자가 필요해보여요~

ZZBAE replied:
접근제어자 붙여주었습니다!  :)

protocorn93 commented on OpenMarket/OpenMarket/View/MarketCollectionViewCell.swift:
```diff
@@ -0,0 +1,92 @@
+            case .failure(_):
+                print("서버 통신 실패")
+            }
+        }
```
이렇게 네트워크를 통해 이미지를 다운로드받는 기능을 갖는 커스텀 UIImageView를 만들어보시는 것도 좋을 것 같아요.

Judy-999 replied:
`UIImageView`를 상속받는 `SessionImageView`를 생성해 **MarketGridCollectionViewCell**과 **MarketListCollectionViewCell**이 `UIImageView` 대신에  `SessionImageView`를 갖도록 했습니다.

`SessionImageView`에 세션을 통해 이미지를 다운받는 `configureImage` 메서드를 생성해 이미지를 할당하도록 했습니다.
<br>


```swift
open class UIImageView : UIView { ... }
```
그런데 `UIImageView`도 `UIView`를 상속받은 뷰의 일종인데 아래와 같은 `init`들은 필요하지 않는 것일까요?
` override init(frame: CGRect)`
`required init?(coder: NSCoder)`

protocorn93 replied:
필요하지요~! 이미지뷰를 위한 생성자가 별도로 있는 것일 뿐이에요. 그 생성자 안에서 말씀해주신 두 가지 생성자가 별도로 호출될거에요!

protocorn93 commented on OpenMarket/OpenMarket/View/MarketCollectionViewCell.swift:
```diff
@@ -0,0 +1,92 @@
+        label.font = .preferredFont(forTextStyle: .body)
+        label.textColor = .systemGray
+        return label
+    }()
```
역시 마찬가지로 접근제어자가 필요할 것 같아요.

protocorn93 replied:
상속을 해서 사용하려다보니 이렇게 열어둔 것 같은데, 셀 같은 경우에도 전역적으로 공통적인 기능을 갖는 것이 아니라면 단순히 생김새가 비슷하다고 해서 상속을 사용하는 것은 지양하는 것이 좋아요. 저는 두 셀을 별도로 가져가는게 좋을 것 같아요

이렇게 상속을 위해 열어둔 것도 문제가될 뿐 더러 상속이 지금 당장은 문제를 해결한 것 처럼 보이지만 부모 클래스와 자식 클래스가 서로 강하게 연결되어 있기에 변경에 취약하며 코드가 덕지덕지 붙는 결과를 초래할 수 있어요~ 

이에 대해서는 상속(inheritance) vs 합성(composition)에 대한 글들을 찾아 학습해보시면 도움이 되실 것 같아요.

ZZBAE replied:
생김새가 비슷하고 겹치는게 많아 보여 상위 MarketViewCell을 만들어서 상속을 사용해주었던 것인데, 전역적으로 공통적인 기능을 갖지는 않기 때문에 각각의 Cell에 다시 원래대로 넣어주었습니다. 부모 클래스와 자식 클래스의 강한 연결이 있다고 생각을 못하였는데 상속과 합성을 공부하면서 앞으로 이런 비슷한 상황에서는 콘이 말씀해주신 것 처럼 지양하도록 하겠습니다 😖

protocorn93 commented on OpenMarket/OpenMarket/ViewController/MarketCollectionViewController.swift:
```diff
@@ -129,13 +130,30 @@ final class MarketCollectionViewController: UICollectionViewController {
         }
     }
     
+    private func configureImage(_ item: Item, _ indexPath: IndexPath, _ cell: MarketCollectionViewCell) {
+        self.sessionManager.receiveData(baseURL: item.productImage) { result in
+            switch result {
+            case .success(let data):
+                guard let imageData = UIImage(data: data) else { return }
+                
+                DispatchQueue.main.async {
+                    guard indexPath == self.collectionView.indexPath(for: cell) else { return }
+                    cell.imageView.image = imageData
+                }
+            case .failure(_):
+                print("서버 통신 실패")
+            }
+        }
+    }
```
별도로 바깥으로 추출하게된 이유가 궁금해요~

Judy-999 replied:
ListCell과 GrideCell의 registraion을 설정하는 클로저 내에서 동일하게 사용되는 코드라 반복을 줄여주기 위해 분리했습니다.
현재는 해당 함수를 `SessionImageView`의 메서드로 변경했습니다

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -0,0 +1,44 @@
+
+class AddProductViewController: UIViewController {
+    private let productView = AddProductView()
+    private lazy var dataSource = [UIImage(named: "바보전구"), UIImage(named: "바보전구"), UIImage(named: "바보전구"), UIImage(named: "바보전구"), UIImage(systemName: "plus")]
```
데이터 소스를 보니까 이미지 추가 셀도 하나의 셀로 가져가려는 것 같아요. 하지마 이 둘은 엄연히 다른 역할이라고 생각해요. 

그렇기 때문에 이미지 추가 버튼 셀과 추가한 이미지를 나타내는 셀 두 가지는 별도로 가져가는 것이 좋을 것 같아요

ZZBAE replied:
이미지 추가 버튼과 이미지 셀을 같이 생각했었는데, 콘 말씀 덕분에 둘의 역할이 다르고 그래서 별도로 분리해줘야한다는걸 깨달았습니다! 밑의 코멘트와 같이 진행해보겠습니다. 🙇🏻‍♀️

protocorn93 commented on OpenMarket/OpenMarket/View/AddProductView.swift:
```diff
@@ -0,0 +1,60 @@
+
+import UIKit
+
+final class AddProductView: UIView {
```
이미지 등록/수정하는 화면 전체를 콜렉션 뷰로 만드는게 어떨까요? 각 입력 form이 하나의 셀이 되겠죠~
보통 이렇게 form 형태으 화면은 그렇게 만들곤해요. 텍스트를 입력한다면 더더욱 그렇게 해주는게 좋아요. 해상도가 작다면 텍스트 필드를 키보드가 가리게될텐데, 이때 content inset을 조정해서 이를 대응할 수도 있구요

ZZBAE replied:
말씀해주신대로 상품을 등록하고 수정하는 화면 전체를 콜렉션 뷰로 만들어보겠습니다. 위에서 코멘트 달아주신 이미지 추가 버튼과 추가한 이미지를 보여주는 셀 두가지를 별도로 나누는 부분과 같이 수정해보도록 하겠습니다. 🙇🏻‍♀️

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -0,0 +1,44 @@
+
+    @objc private func goBackWithUpdate() {
+        self.navigationController?.popViewController(animated: true)
+    }
```
이 버튼들은 `UIBarButtonItem` 들에 대한 액션인가요??

ZZBAE replied:
넵 그렇습니다! 밑의 코멘트와 이어서, BarButton에 대한 네이밍 변경을 해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -18,10 +18,19 @@ class AddProductViewController: UIViewController {
     }
 
+    func configureUI() {
+        let cancelBarButton = UIBarButtonItem(title: "Cancel", style: .plain, target: self, action: #selector(goBack))
+        let doneBarButton = UIBarButtonItem(title: "Done", style: .plain, target: self, action: #selector(goBackWithUpdate))
```
`cancelButtonDidTapped`, `updateButtonDidTapped`가 좀 더 적절할 것 같아요. 제가 말한 네이밍이 아니더라도 `go-` 네이밍은 적절하진 않은 것 같아요~

ZZBAE replied:
go 로 시작하는 네이밍이 별로라고 생각은 했는데, 어떤 것으로 바꾸어줄지 생각이 안나 이렇게 PR을 하게 되었던 것 같습니다ㅜㅜ 네이밍 변경을 해주니까 훨씬 읽기 쉬워졌습니다. 추천해주신대로 `cancelButtonDidTapped`와 `updateButtonDidTapped`로 변경해주었습니다! 👍🏻

protocorn93 commented on OpenMarket/OpenMarket/View/AddProductCollectionViewCell.swift:
```diff
@@ -10,21 +10,21 @@ import UIKit
+        let imageView = UIImageView()
+        imageView.backgroundColor = .systemGray5
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
```
이미지 추가 버튼 셀과 이미지 셀을 같이 사용하다보니 이런 코드 수정이 발생한 것 같네요!

ZZBAE replied:
이미지 추가 버튼 셀과 이미지 셀을 분리하겠습니다!! 🥲

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -47,7 +56,41 @@ extension AddProductViewController: UICollectionViewDataSource {
+            self.present(self.imagePicker, animated: true)
+        }
+        
+        if indexPath.row == 5 {
+            showAlert(title: "사진 등록 불가능", message: "사진은 최대 5장까지 가능합니다.")
+        }
```
이미지 갯수에 대한 검증은 이미지 피커 델리게이트 메서드에서 진행해주는 것이 적절한 것 같아요.

Judy-999 replied:
```swift
 func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) { 
     guard imageParams.count != 5 else {
            picker.dismiss(animated: true, completion: nil)
            showAlert(title: "사진 등록 불가능", message: "사진은 최대 5장까지 가능합니다.")
            return
        }
    // 생략
}
```

이미지가 5개를 넘는지 확인하는 코드를 `imagePickerController` 메서드로 이동시켰습니다!

그런데 위의 메서드는 이미지를 선택하면 실행되는 메서드라 6번째 이미지를 선택 완료했을 때 이미지를 더 넣을 수 없다고 알려주게 되는데, 5번째 사진은 아예 선택할 수 없게 하는게 맞지 않나 싶은데 콘의 생각은 어떠신지 궁금합니다! 🤔

protocorn93 replied:
저는 그렇게 갯수 제한을 하는게 맞지 않나 싶어요! 나중에는 Photos 프레임워크를 사용해서 직접 사진 콜렉션 화면도 만들어보시면 좋을 것 같아요. 다중 선택, 선택 취소도 구현해보시면 도움이 많이 될 것 같아요!

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -47,7 +56,41 @@ extension AddProductViewController: UICollectionViewDataSource {
+    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
+        var selectedImage: UIImage? = nil
+        
+        if let newImage = info[UIImagePickerController.InfoKey.editedImage] as? UIImage {
+            selectedImage = newImage
+        } else if let newImage = info[UIImagePickerController.InfoKey.originalImage] as? UIImage {
+            selectedImage = newImage
+        }
```
이 둘의 차이점에 대해 설명해주실 수 있으신가요~?

ZZBAE replied:
아무 편집없이 그냥 선택된 이미지면 `originalImage`를 받아오고, 편집이 되었으면 `editedImage`로 받아올 수 있습니다. 이미지를 편집했을 때 편집된 이미지를  if let으로 editedImage를 먼저 받아올 수 있게 해주었고, 편집된 것이 없으면 `originalImage` 를 받아오게 해주었습니다.

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -11,20 +11,23 @@ class AddProductViewController: UIViewController {
     private let imagePicker = UIImagePickerController()
-
+    
+    lazy var viewConstraint = productView.entireStackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -260)
```
앞의 코멘트와 같이 전체를 콜렉션 뷰로 만들고 content inset을 조정하시는게 좀 더 바람직한 방법일 것 같아요. 

그리고 keyboard show/hide 이벤트를 구독해서 처리해보시는 걸 권장해드려요. 해당 이벤트를 통해 키보드의 높이, 키보드 show/hide 애니메이션 duration도 알 수 있으니 좀 더 섬세한 제어가 가능할거에요

protocorn93 commented on OpenMarket/OpenMarket/ViewController/AddProductViewController.swift:
```diff
@@ -37,17 +36,70 @@ class AddProductViewController: UIViewController {
     
     @objc private func goBackWithUpdate() {
+        let sessionManager = URLSessionManager(session: URLSession.shared)
+        guard let param = productView.receiveParam() else { return }
```
네트워크 요청을 위한 데이터를 조합하는 과정도 별도 객체에게 위임하는 것이 좋을 것 같아요.

ZZBAE replied:
서버에 request로 post할 데이터들의 조합하는 기능을 분리시켜 `ParamManager`라는 객체를 생성시켜 위임해주었습니다! 🫡

protocorn93 reviewed (changes requested):
코멘트가 늦어져서 죄송해요 😭 이미지 업로드라는 것이 절대 쉬운 피처가 아닌데 두 분 모두 고생하셨어요! 

# 궁금한 점

> 1. 텍스트뷰 자동 스크롤 - 키보드 사이즈

이에 대해서는 코멘트도 남겨놓았지만 전체를 콜렉션 뷰로 하고 form 하나 하나를 별도 셀로 가져가는 것이 더 바람직해요. 실제로 이런 형태의 화면은 대부분 콜렉션뷰로 만들곤해요.

그리고 `UIResponder.keyboardWillShowNotification,` `UIResponder.keyboardHideShowNotification`를 observe 해서 키보드에 대한 처리를 해주시면 될 것 같아요. 이에 대해서도 역시 코멘트 남겨놓았아요. 

> 2. 사진을 5개 넣으면 POST 되지 않는 문제

이 부분에 대해서는 다양한 원인이 있을 수 있는데, 저도 조금 더 살펴보고 답을 찾으면 말씀드려볼게요! 

> 3. 사진 타입 구하기

이에 대해서는 여러가지 방법이 있을 수 있는데, https://stackoverflow.com/questions/4147311/finding-image-type-from-nsdata-or-uiimage/40317201#40317201 이 방법도 주로 사용되는 방법으로 알고 있어요~ 

> 4. 사진 추가 셀의 이미지 (+)

이에 대해서도 역시 코멘트를 남겨놓았는데요~ 기본적으로 이미지 추가 버튼 셀과 이미지 셀을 분리하고 버튼으로 가져가는 것이 적절해보여요. 


# 고민한 점

> 1. 스크롤 시 상품에 맞지 않는 이미지가 나오는 문제 해결

셀에 이미지를 보여줄 때는 다양한 이슈가 발생할 수 있는데 이 역시 그 중 하나라고 보여져요. 사용하신 방법도 하나의 방법이겠죠. 추가적으로 캐싱을 함께 구현해보시면 좋을 것 같아요. 매번 셀이 보여질 때마다 호출하는 것은 비효율적이라고 판단이 되네요!

Judy-999 committed:
refactor: 여러 장의 사진이 post되지 않는 문제 해결

- boundary와 dispostion을 사진 마다 추가되도록 위치를 for문 안으로 이동
- 테스트용 이미지 파일 삭제
- 옵셔널바인딩 방식을 extension을 통해 사용하던 메서드를 guard let으로 변경

Judy-999 committed:
Merge branch: 리드미 업데이트한 브랜치와 머지

ZZBAE committed:
refactor: List와 Grid의 상속 관계 수정 및 메서드 정리

- MarketCollectionViewCell의 삭제
- MarketGridCollectionViewCell과 MarketListCollectionViewCell의 상속을 제거하고 별도의 셀로 가져가게끔 수정
- configureImage 메서드를 분리하여 GridDataSource와 ListDataSource에서 각각 실행하도록 변경

ZZBAE committed:
refactor: 버튼액션 메서드 네이밍 변경 및 Delegate 활용 방법 변경

- 이미지 갯수에 대한 검증을 UIImagePickerControllerDelegate 메서드를 사용하게끔 변경
- BarButton의 액션인 메서드의 네이밍 변경

Judy-999 committed:
refactor: SessionImageView 추가

- SessionImageView : UIImageView를 커스텀한 뷰로 세션을 통해 이미지를 다운받음
- List와 Grid Cell이 UIImageView 대신 SessionImageView를 갖도록 변경

Judy-999 committed:
refactor: param으로 post할 데이터를 조합하는 ParamManager 생성

ZZBAE committed:
refactor: 이미지 파일의 타입 구별 기능 추가

- data타입을 extension해서 post할 이미지 파일의 타입을 구별해주는 fileExtension 추가

ZZBAE committed:
refactor: ImageCacheManager 이미지의 Cache 구현

- 이미지가 호출될 때마다 다운받아오게 하지 않고, cache를 사용하여 이미지 임시저장 후 바로 가져오는 기능 구현

ZZBAE committed:
refactor: 접근제어자 추가 및 무의미한 파라미터 삭제

- imageView에 접근제어자 추가
- configureCell 메서드의 spacingType 파라미터 제거 및 각각의 셀에 넣어줄 띄어쓰기 타입 string으로 추가

Judy-999 commented:
안녕하세요 콘 이번에도 세심한 리뷰 감사드립니다 😊

위에 댓글 남긴 것 외에 해결한 문제들입니다.
<br>

###  1. 사진을 5개 넣으면 POST 되지 않는 문제
body를 생성하는 코드를 다시 살펴본 결과 이미지 파트를 넣는 부분에 문제가 있었습니다.
```swift
--Boundary-UUID
Content-Disposition:form-data; name="images"; filename="사진이름1.jpeg"
Content-Type: image/jpeg

이미지 데이터1
--Boundary-UUID
Content-Disposition:form-data; name="images"; filename="사진이름2.jpeg"
Content-Type: image/jpeg

이미지 데이터2
--Boundary-UUID--
```
위와 같은 형식으로 만들어져야 했는데,
```swift
--Boundary-UUID
Content-Disposition:form-data; name="images"; filename="사진이름1.jpeg"
Content-Type: image/jpeg

이미지 데이터1		         // 밑에 Boundary가 없음
; filename="사진이름2.jpeg"	// Content-Disposition:form-data; name="images" 없이 ;filename 부터 시작
Content-Type: image/jpeg

이미지 데이터2
--Boundary-UUID--
```
이처럼 파트의 헤더부분이 제대로 들어가지 않아 발생했던 문제였습니다. 
`Boundary`와 ` Content-Disposition`를 넣어주니 해결되었습니다!
<br>

### 2. 이미지 타입
알려주신 사이트를 참고하여 "jpeg"로 고정하지 않고 사진 별로 타입을 구별해서 넣어주도록 했습니다.
